feat(appsync): support S3 location for GraphQL schema

BryanPan342 opened the pull request:
Add support for GraphQL Schema to define from an s3 location. For convenience, use the `appsync.Schema.fromS3Location` to specify the s3 location representing your schema.

<details>

<summary> Example </summary>

```ts
const bucket = new s3.Bucket(stack, 'Bucket');

const deploy = new s3deploy.BucketDeployment(stack, 'Deployment', {
  sources: [s3deploy.Source.asset(path.join(__dirname, 'schemas'))],
  destinationBucket: bucket,
});

const api = new appsync.GraphqlApi(stack, 'Api', {
  name: 'integ-test-s3',
  schema: appsync.Schema.fromS3Location({
    bucket,
    key: 'appsync.test.graphql',
  }),
});

api.node.addDependency(deploy);
```

</details>

**BREAKING CHANGES**: `appsync.Schema` changed to an abstract class
* **appsync**: No longer supporting `new appsync.Schema()`, instead use the static methods `appsync.Schema.fromCode()`, `appsync.Schema.fromAsset()`, `appsync.Schema.fromS3()`

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 6fb32765a10ceaf528e620ffc34cd86e3de89b78
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4215fb8a-323c-4ebf-aee9-ba47a86e019c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: e3a437e06879a690cd17d2465b6064e7adf8f518
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=43b69656-1c28-4bde-9adf-973ded53ea4f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8215825d24ef0078bbf83504015adc45602ccd59
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=93eaea39-3b9e-4a88-98c8-99e6fc5d6b42%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

MrArnoldPalmer commented:
@BryanPan342 we have had this one assigned to you for awhile. Are you still expecting changes or did we just forget to move it to myself or @shivlaks?

BryanPan342 commented:
@MrArnoldPalmer  honeslty i forgot but im pretty good with it as it is. 

i think we had some apprehensions with the underlying implications of assets.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 82e6d1795faf51325ecffce551d94efc67630a92
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b9af3488-c012-447d-a49c-d9a7c80d129c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 4a2736db13e09190b1b566a9ad249db920738e8b
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ca24f854-65b3-4298-82e8-29995e228764%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appsync/README.md:
````diff
@@ -327,6 +329,18 @@ const api = appsync.GraphqlApi(stack, 'api', {
+```ts
+const api = appsync.GraphqlApi(stack, 'api', {
+  name: 'myApi',
+  schema: appsync.Schema.fromS3Location('s3://example-url'),
````
Can we show a real example here? For an actual file inside a Bucket?

skinny85 commented on packages/@aws-cdk/aws-appsync/lib/schema.ts:
```diff
@@ -34,13 +43,23 @@ export class Schema {
+   * @returns `Schema` with immutable schema defintion
+   * @param s3Location the s3 location of the schema
+   */
+  public static fromS3Location(s3Location: string): Schema {
```
I wonder if just a string is a good API here. We often represent S3 locations as pairs of `IBucket` and object key.

We also have the [`Location` interface in the S3 module](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-s3.Location.html). Maybe that's a better abstraction than just a lonely string?

BryanPan342 replied:
hm totally right.. ill use the IBucket and object key method alongside `s3UrlForObject`

skinny85 commented on packages/@aws-cdk/aws-appsync/lib/schema.ts:
```diff
@@ -59,9 +78,16 @@ export class Schema {
   private types: IIntermediateType[];
 
   public constructor(options?: SchemaOptions) {
+    if (options?.filePath && options?.s3Location) {
```
The fact that you have to write this validation suggests to me this API could be better.

Can we actually change this API here a little bit?

1. Make `Schema` abstract.
2. Create a subclass of `Schema`, `SchemaImpl`, that's module-private, and contains all of the logic here. It extends `Schema`, and takes `SchemaOptions` in its constructor argument. `SchemaOptions` should now also be module-private.
3. All of the static factory methods declare that they return `Schema` in their signature, but return `SchemaImpl` in their bodies.

This way, you will be in control of the invariants of `SchemaImpl`, and you won't have to write this defensive code. In fact, you can have a separate subclass of `Schema` for each of the static factory methods!

BryanPan342 replied:
I'm all for this change, but it would be a breaking change because we might have users using the `Schema` api using `new Schema()`. Are we comfortable with introducing this breaking change?

skinny85 replied:
Yep, [`appsync` is still experimental](https://github.com/aws/aws-cdk/blob/1eb6de7913a27874a2c8c382900d6976fe7f0564/packages/%40aws-cdk/aws-appsync/package.json#L124-L125).

BryanPan342 replied:
Sweet, ill get it done today

skinny85 commented on packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:
```diff
@@ -32,6 +33,16 @@ beforeEach(() => {
   stack = new cdk.Stack();
 });
 
+test('appsync fails when both file path and s3 location are configured', () => {
```
I would say the schema fails creation, "appsync fails" is kind of a weird statement 😃.

skinny85 commented on packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:
```diff
@@ -272,4 +283,108 @@ describe('testing schema definition mode `file`', () => {
+    // WHEN
+    new appsync.GraphqlApi(stack, 'API', {
+      name: 'demo',
+      schema: appsync.Schema.fromS3Location(schemaAsset.s3ObjectUrl),
```
This is a weird test.. we have a separate `fromAsset()` static factory method for `Schema`, so why also use Asset here...?

skinny85 commented on packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:
```diff
@@ -272,4 +283,108 @@ describe('testing schema definition mode `file`', () => {
+    });
+  });
+
+  test('definition mode `s3` errors when addType for object is called', () => {
+    // WHEN
+    const api = new appsync.GraphqlApi(stack, 'API', {
+      name: 'demo',
+      schema: appsync.Schema.fromS3Location('garbage'),
+    });
+
+    // THEN
+    expect(() => {
+      api.addType(new appsync.ObjectType('blah', {
+        definition: { fail: t.id },
+      }));
+    }).toThrowError('API cannot add type because schema definition mode is not configured as CODE.');
+  });
+
+  test('definition mode `s3` errors when addType for interface is called', () => {
+    // WHEN
+    const api = new appsync.GraphqlApi(stack, 'API', {
+      name: 'demo',
+      schema: appsync.Schema.fromS3Location('garbage'),
+    });
+
+    // THEN
+    expect(() => {
+      api.addType(new appsync.InterfaceType('blah', {
+        definition: { fail: t.id },
+      }));
+    }).toThrowError('API cannot add type because schema definition mode is not configured as CODE.');
+  });
```
These 2 tests are exactly the same, no?

BryanPan342 replied:
An interface type is technically different than an object type but the validation is the same.. ill delete the interface one and rename the test to test against `IntermediateType`

skinny85 replied:
Let me ask this 🙂.

Is there anything that you can pass to `. addType()` there that would _not_ throw an exception?

BryanPan342 replied:
ah yes that is very true

skinny85 commented on packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:
```diff
@@ -272,4 +283,108 @@ describe('testing schema definition mode `file`', () => {
+
+  test('definition mode `s3` errors when addMutation is called', () => {
+    // WHEN
+    const api = new appsync.GraphqlApi(stack, 'API', {
+      name: 'demo',
+      schema: appsync.Schema.fromS3Location('garbage'),
+    });
```
These lines are repeated in like 5 tests in a row. Can you introduce a `describe()` block with a `beforeEach()` that gets rid of this duplication?

skinny85 reviewed (changes requested):
Looks great! A couple of questions/suggestions around the API experience here, but in general super solid!

BryanPan342 commented:
@skinny85 updated to reflect the new model

I did a reorg as well so sorry for the big PR but here is a quick summary:

* Moved `private.ts` to a `private/` directory to hold both MappingTemplates and Schema
* Removed dependency w/ `@aws-cdk/aws-s3-assets` and instead rely on `@aws-cdk/aws-s3` and `@aws-cdk/aws-s3-deployment`
* Moved all the testing `graphql` files under the `test/schema/` directory because I needed to upload to s3 and the `test/` dir was getting out of hand

I also did a manual integ test to make sure the schema is accessible by the GraphQL Api!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 7da84de4d2ee64918108416e3a34f92c407353fa
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fde39111-6ce4-4a82-913c-63d95706268e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: aa02a5f9ae50ca3aea473138ef21ab77108d927d
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=18f1b7ad-f2f0-4355-bf8f-f25b9203aebd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: e2dec6f7d9153754db7dfbaa612933e1209bf493
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=57d94aed-dff1-4274-ae63-a460c2c869b7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appsync/lib/schema.ts:
```diff
@@ -1,219 +1,55 @@
  * code-first.
  */
-export class Schema {
+export abstract class Schema extends SchemaImpl {
```
This is the wrong way around. `SchemaImpl` should extend `Schema`.

skinny85 commented on packages/@aws-cdk/aws-appsync/lib/schema.ts:
```diff
@@ -1,219 +1,55 @@
-    }
-    this.types.push(type);
-    return type;
+  public static fromS3Location(s3Location: S3Location): Schema {
```
Make these positional arguments instead, the interface doesn't buy us much here.

skinny85 commented on packages/@aws-cdk/aws-appsync/lib/private/mapping-template.ts:
```diff
@@ -0,0 +1,113 @@
+/**
```
Why is this file called `mapping-template.ts`? There's no such class/function in this file.

BryanPan342 replied:
It holds helper functions for the mapping-templates!

BryanPan342 replied:
we could call it `mapping-templates-utils.ts`?

skinny85 replied:
How about something like `conditions.ts`, since those seem to be the classes present here?

BryanPan342 replied:
Okay im good with that

skinny85 commented on packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:
```diff
@@ -0,0 +1,322 @@
+  /**
+   * Add type to the schema
+   */
+  public addType(_type: IIntermediateType): IIntermediateType {
+    throw new Error('API cannot add type because schema definition mode is not configured as CODE.');
+  }
```
I believe you're doing this backwards.

All these methods should be `asbtract` in `Schema`, and then `SchemaImpl` should actually implement them.

skinny85 commented on packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:
```diff
@@ -0,0 +1,322 @@
+    if (!this._schema) {
+      this._schema = new CfnGraphQLSchema(api, 'Schema', {
+        apiId: api.apiId,
+        definitionS3Location: Lazy.string({
```
Why the `Lazy` here?

skinny85 commented on packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:
```diff
@@ -0,0 +1,322 @@
+    super();
+    this.definition = readFileSync(filePath).toString('utf-8');
+  }
+}
```
Sorry, maybe I'm missing something.

Why does the `AssetSchema` read the file directly from disk? Doesn't this mean the Asset is not actually needed? (Assets are actually deployed to S3, but it seems like this is not needed here)

BryanPan342 replied:
So originally the way AppSync was created was that we supported reading the `graphql` file. So in creating the static function that read a local file, we decided to follow the same scheme as Lambda `Code.fromAsset(...)`. Im guessing Asset is a misleading word?

skinny85 replied:
A little bit, in my opinion. "Asset" is a domain concept from the CDK (see [here](https://docs.aws.amazon.com/cdk/latest/guide/assets.html)).

I personally would have called that method `fromFile()`.

But let's maybe leave this one for now. It's kind of a separate issue. So I won't block the PR on it.

BryanPan342 replied:
Hm yeah okay I can live with that. If it makes your life easier, I can also split this PR into two to lower the file count!

skinny85 commented on packages/@aws-cdk/aws-appsync/lib/private/schema.ts:
```diff
@@ -0,0 +1,115 @@
+import { AuthorizationType } from '../graphqlapi';
```
Can we call this file something like `schema-shape-addition.ts`? The current name really doesn't say anything.

skinny85 commented on packages/@aws-cdk/aws-appsync/lib/schema.ts:
```diff
@@ -1,219 +1,55 @@
-    };
-    this.query.addField({ fieldName, field });
-    return this.query;
+    return new AssetSchemaImpl(filePath);
```
I think you can cut the `Impl` from the name, and just call this `AssetSchema`.

skinny85 commented on packages/@aws-cdk/aws-appsync/lib/schema.ts:
```diff
@@ -1,219 +1,55 @@
-    this.types.push(type);
-    return type;
+  public static fromS3Location(s3Location: S3Location): Schema {
+    return new S3SchemaImpl(s3Location);
```
Again, let's just call this `S3Schema`.

skinny85 commented on packages/@aws-cdk/aws-appsync/lib/schema.ts:
```diff
@@ -1,219 +1,55 @@
-        .filter((field) => field != ''),
-    }) + '\n';
+  public static fromCode(): Schema {
+    return new CodeSchemaImpl();
```
Again, let's just call this `CodeSchema`.

skinny85 commented on packages/@aws-cdk/aws-appsync/test/appsync-schema.test.ts:
```diff
@@ -222,54 +215,110 @@ describe('testing schema definition mode `file`', () => {
+      }).toThrowError('Unable to add query. Schema definition mode must be CODE. Received: S3');
+    });
+
+    test('definition mode `s3` errors when addMutation is called', () => {
```
The descriptions of these tests could be improved. Remember that the parent descriptions are concatenated with the children. They should form proper sentences.

Something like:

```ts
  describe('API with Schema from S3', () => {
   // ...

  test('throws an exception when addQuery() is called on it', () => {
    // ...
```

skinny85 commented on packages/@aws-cdk/aws-appsync/test/integ.graphql-s3.ts:
```diff
@@ -0,0 +1,61 @@
+  }),
+});
+
+api.node.addDependency(deploy);
```
Really don't like the fact that you have to call this yourself.

Should we have a `Schema` variant that takes in the S3 deployment type, that handles this for you?

BryanPan342 replied:
Hm.. i think this is residual code that I had before when I was trying to get the bucket token to resolve

BryanPan342 replied:
Actually nevermind, I retract that previous statement. Without the dependency between the deployment and schema, CDK doesn't know to that it needs to deploy the Schema after running the deployment. 

We currently have an API for the api called `addSchemaDependency` for the `GraphqlApi` construct.

If we want to make this a smoother process we could do the following:

```ts
export class S3Schema extends SchemaImpl {
  ...
  public constructor(bucket: IBucket, key: string, deployment?: BucketDeployment) {
     ...
     this.deployment = deployment.
  }

  public bind(api) {
    ...
    this.schema.node.addDependency(this.deployment);
  }
```

skinny85 replied:
This needs to be a separate static factory option that takes in a `BucketDeployment` (not both an `IBucket` and a `BucketDeployment`, because `BucketDeployment` already has an `IBucket` defined).

BryanPan342 replied:
ah yes that makes sense

BryanPan342 replied:
Should that change occur in another PR? And keep this one focused on just `fromS3Location`? In fact, I wouldn't mind changing the static function name to `fromBucket`

skinny85 replied:
Sure.

skinny85 reviewed (changes requested):
The PR is pretty tough to review, but I've seen worse 😉.

A few comments before this can be merged.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 9bdde0901ef608d4bc0e4b52a9335ad295698e32
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=807192a3-ddb9-483a-a419-f78912ec2778%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

BryanPan342 commented on packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:
```diff
@@ -0,0 +1,322 @@
+ *
+ * @internal
+ */
+export class SchemaImpl {
```
ahhh i see thats what i was getting wrong... I think i misread your message hence the circular dependency stuff 🥲

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 2821d7ef99beda7c32e3895e62b6d2e6e597e0bc
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c66d9323-d69b-47ca-9dd6-797762361ccb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: f3a6df2c1e57adfcf8871385c86dbeeb6015778d
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c97d4c60-98df-4c0b-ad16-75c4421402b2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: abc0924b4692ac65aa6316a683beb3b7d96a7f5d
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3a911585-4807-451c-b33d-ef3df4e893f2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: da1ae62779aaa7063dc83e1fae2276647a4de9c4
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=df5f1eef-7fc3-4d5b-bbe3-3896e235cd22%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: f7e36dfc29a478bb757f9b96673107f3089700fe
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cfe40293-4da8-4024-bf53-f688361b9af9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 92cff3919a46044bd9d6acc60127a8e51cfe2f8a
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8952a3f5-946d-4cdb-9d09-d262724abb9c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 reviewed (changes requested):
Looks good, but I think I came up with a way to make `Schema` great 🙂. Let me know what you think!

BryanPan342 commented on packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:
```diff
@@ -0,0 +1,326 @@
+    * to the stack.
+    */
+  public bind(api: GraphqlApi): CfnGraphQLSchema {
+    if (!this._schema) {
```
I dont want to create two schemas. So the conditional checks if it was created already.

skinny85 replied:
Why?

BryanPan342 replied:
I mean.. if there is not point in it, why take up the memory to run it again? And I think the schema should be the same no? Unless there is some use case where you want to make multiple stacks that each have a schema that builds on top of itself?

skinny85 replied:
Look at this code:

```ts
const schema = appsync.Schema.fromCode();
new GraphqlApi(this, 'Api1', { schema, });
new GraphqlApi(this, 'Api2', { schema, });
```

Because of  the caching that you do, the second `Api` does not have a Schema at all (its call to `Schema.bind()` returns the existing `CfnGraphQLSchema` with `apiId` filled with the ID of the first API).

I cant imagine this is the correct behavior.

BryanPan342 commented on packages/@aws-cdk/aws-appsync/lib/schema.ts:
```diff
@@ -1,134 +1,74 @@
+  /**
+   * The definition for this schema
+   */
+  public readonly abstract definition: string;
```
it existed before, so i kept it as to avoid breaking changes

skinny85 replied:
Let's kill it.

BryanPan342 commented on packages/@aws-cdk/aws-appsync/lib/private/schema-implementation.ts:
```diff
@@ -0,0 +1,326 @@
+    * Add type to the schema
+    */
+  public addType(_type: IIntermediateType): IIntermediateType {
+    throw new Error('API cannot add type because schema definition mode is not configured as CODE.');
```
wow.. i honestly did not know you could do this 😱  my mind is actually blown what

BryanPan342 replied:
question, how would a user add this to the GraphqlApi? Because the Graphql API has the option of adding queries, and mutations to the schema as well!

https://github.com/aws/aws-cdk/blob/6b639bee925a9e552c405fde9d221d515e6023c3/packages/%40aws-cdk/aws-appsync/lib/graphqlapi.ts#L577-L643

If i set the type for schema to be a `Schema` it is immutable and thus cannot be targeted by the API. I guess what we **could** do is remove that feature entirely? Prevents new users from tripping and trying to add a Query if they are importing their schema locally.

skinny85 replied:
Maybe keep the API for now, but move the error code to `GraphqlApi` (basically, each of the methods does an `instanceof` check, and throws if the `schema` is not a `MutableSchema`)?

TheRealAmazonKendra commented:
This PR has been deemed to be abandoned, and will be closed. Please create a new PR for these changes if you think this decision has been made in error.
